### PR TITLE
Prevent extreme duplicate SQL queries

### PR DESCRIPTION
**Pull request by JeroenSormani**
When a discount is applied to the cart it will call the `edd_get_discounts` a lot of times. This function calls the WP_Query which causes duplicate queries. When there's only one product in the cart its 'only' 82 queries instead of ~32. When there's more it will multiply quickly and with a test of 9 products in the cart it will send 500+ duplicate queries to the DB.

https://cl.ly/262J0O1i0X2V

--------

I've also investigated the roots of it being called that many times; this is something else that I'd like to discuss sometime soon. The results of some quick tests seem very promising for improvements in the checkout page speed. 

Before going from 2.2~2.3 seconds (with pretty regular peeks of ~5+ seconds); going to 1.4-1.6 seconds without those spikes.

### CODE REVIEW

File: includes/discount-functions.php [hunk 1]
```diff
@@ -32,19 +32,29 @@ function edd_get_discounts( $args = array() ) {
 
 	$args = wp_parse_args( $args, $defaults );
 
-	$discounts = get_posts( $args );
+	$discounts_hash = md5( json_encode( $args ) );
+	if ( ! $discounts = wp_cache_get( $discounts_hash, 'edd-discounts' ) ) {
+		$discounts = get_posts( $args );
+		wp_cache_set( $discounts_hash, $discounts, 'edd-discounts' );
```
Comment (sunnyratilal) [inline includes/discount-functions.php]: An expiry needs to be set to `3600` otherwise it default to `0` which means no expiration.

File: includes/discount-functions.php [hunk 2]
```diff
@@ -32,19 +32,29 @@ function edd_get_discounts( $args = array() ) {
...
 		$args['meta_key']     = '_edd_discount_code';
 		$args['meta_value']   = $args['s'];
 		$args['meta_compare'] = 'LIKE';
 		unset( $args['s'] );
-		$discounts = get_posts( $args );
+
+		$discounts_hash = md5( json_encode( $args ) );
+		if ( ! $discounts = wp_cache_get( $discounts_hash, 'edd-discounts' ) ) {
+			$discounts = get_posts( $args );
+			wp_cache_set( $discounts_hash, $discounts, 'edd-discounts' );
```
Comment (sunnyratilal) [inline includes/discount-functions.php]: Same here - an expiry needs to be set to `3600` otherwise it default to `0` which means no expiration.

File: includes/discount-functions.php [hunk 3]
```diff
@@ -32,19 +32,29 @@ function edd_get_discounts( $args = array() ) {
...
 	if( ! $discounts && ! empty( $args['s'] ) ) {
-		// If no discounts are found and we are searching, re-query with a meta key to find discounts by code
 		$args['meta_key']     = '_edd_discount_code';
 		$args['meta_value']   = $args['s'];
 		$args['meta_compare'] = 'LIKE';
 		unset( $args['s'] );
-		$discounts = get_posts( $args );
+
+		$discounts_hash = md5( json_encode( $args ) );
+		if ( ! $discounts = wp_cache_get( $discounts_hash, 'edd-discounts' ) ) {
```
Comment (sunnyratilal) [inline includes/discount-functions.php]: In order to maintain compatibility with PHP 5.3, please declare and call the function outside the `if` statement and then reference the variable.

i.e.
```php
$discounts = wp_cache_get( $discounts_hash, 'edd-discounts' )
if ( false === $discounts ) {
```

File: includes/discount-functions.php [hunk 4]
```diff
@@ -32,19 +32,29 @@ function edd_get_discounts( $args = array() ) {
 
 	$args = wp_parse_args( $args, $defaults );
 
-	$discounts = get_posts( $args );
+	$discounts_hash = md5( json_encode( $args ) );
+	if ( ! $discounts = wp_cache_get( $discounts_hash, 'edd-discounts' ) ) {
```
Comment (sunnyratilal) [inline includes/discount-functions.php]: In order to maintain compatibility with PHP 5.3, please declare and call the function outside the `if` statement and then reference the variable.

i.e.
```php
$discounts = wp_cache_get( $discounts_hash, 'edd-discounts' )
if ( false === $discounts ) {
```

Comment (JeroenSormani) [inline includes/discount-functions.php]: `if ( ! $discounts = wp_cache_get( $discounts_hash, 'edd-discounts' ) ) {` 
is valid for php 5.3 (verified), but I can understand if this is not a preferred method. Can you confirm if this needs changing either way?

File: includes/discount-functions.php (hunk 2 above)
Comment (JeroenSormani) [inline includes/discount-functions.php]: Thanks for spotting that, fixed it!

File: includes/discount-functions.php (hunk 4 above)
Comment (sunnyratilal) [inline includes/discount-functions.php]: Yes please - it does need changing

Comment (JeroenSormani) [inline includes/discount-functions.php]: Just pushed the change :-)

File: includes/discount-functions.php [hunk 5]
```diff
@@ -33,7 +33,8 @@ function edd_get_discounts( $args = array() ) {
 	$args = wp_parse_args( $args, $defaults );
 
 	$discounts_hash = md5( json_encode( $args ) );
-	if ( ! $discounts = wp_cache_get( $discounts_hash, 'edd-discounts' ) ) {
+	$discounts = $discounts = wp_cache_get( $discounts_hash, 'edd-discounts' );
```
Comment (sunnyratilal) [inline includes/discount-functions.php]: Needs changing to `$discounts = wp_cache_get( $discounts_hash, 'edd-discounts' );`

Comment (JeroenSormani) [reply includes/discount-functions.php]: Oops! Fixed that.

**Comment by pippinsplugins** [comment]
I'd like to see a few unit added for `edd_get_discounts()` before this merged in.

**Comment by cklosowski** [comment]
If we're doing this, we need cache busting on adding/updating a discount.

**Comment by cklosowski** [comment]
Actually i'm going to pull this into an issue branch and work up some unit tests on it as well.

**Comment by cklosowski** [comment]
Issue #6039 created, closing this. your commit history will remain in tact @JeroenSormani
